joblib/numpy_pickle_utils.py

```diff
@@ -147,20 +151,42 @@ def _read_fileobject(fileobj, filename, mmap_mode=None):
     """
     # Detect if the fileobj contains compressed data.
     compressor = _detect_compressor(fileobj)
+    if isinstance(fileobj, tuple(_COMPRESSOR_OBJS)):
+        compressor = fileobj.__class__.__name__
     if compressor == 'compat':
+        # Compatibility with old pickle mode: simply return the input
+        # filename "as-is" and let the compatibility function be called by the
+        # caller.
         warnings.warn("The file '%s' has been generated with a joblib "
                       "version less than 0.10. "
                       "Please regenerate this pickle file." % filename,
                       DeprecationWarning, stacklevel=2)
         yield filename
     else:
-        if compressor in _COMPRESSORS and mmap_mode is not None:
+        # Checking if incompatible load parameters with the type of file:
+        # mmap_mode cannot be used with compressed file or in memory buffers
+        # such as io.BytesIO.
+        if ((compressor in _COMPRESSORS or
+                isinstance(fileobj, tuple(_COMPRESSOR_OBJS))) and
+                mmap_mode is not None):
             warnings.warn('File "%(filename)s" is compressed using '
                           '"%(compressor)s" which is not compatible with '
-                          'mmap_mode "%(mmap_mode)s" flag passed.'
-                          % locals(), DeprecationWarning, stacklevel=2)
-
-        if compressor == 'zlib':
+                          'mmap_mode "%(mmap_mode)s" flag passed. mmap_mode '
+                          'option will be ignored.'
+                          % locals(), Warning, stacklevel=2)
```

Comment:
For generic warnings, probably better not to specify anything here:

```
warnings.warn(message, stacklevel=2)
```

It will default to UserWarning.